src/CachePlugin.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Allow for caching a response.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class CachePlugin implements Plugin
+{
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $pool;
+
+    /**
+     * Default time to store object in cache. This value is used if CachePlugin::respectCacheHeaders is false or
+     * if cache headers are missing.
+     *
+     * @var int
+     */
+    private $defaultTtl;
+
+    /**
+     * Look at the cache headers to know how long this response is going to be cached.
+     *
+     * @var bool
+     */
+    private $respectCacheHeaders;
+
+    /**
+     * @param CacheItemPoolInterface $pool
+     * @param int                    $defaultTtl
+     * @param bool                   $respectCacheHeaders
+     */
+    public function __construct(CacheItemPoolInterface $pool, $defaultTtl = 60, $respectCacheHeaders = true)
+    {
+        $this->pool = $pool;
+        $this->defaultTtl = $defaultTtl;
+        $this->respectCacheHeaders = $respectCacheHeaders;
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param callable         $next
+     * @param callable         $first
+     *
+     * @return mixed
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        if (!$this->isCacheableRequest($request)) {
+            return $next($request);
+        }
+
+        // If we can cache the request
+        $key = $this->createCacheKey($request);
+        $cacheItem = $this->pool->getItem($key);
+        if ($cacheItem->exists()) {
+            // return cached response
+            return $cacheItem->get();
+        }
+
+        return $next($request)->then(function (ResponseInterface $response) use ($key) {
+            if ($this->isCacheableResponse($response)) {
+                $cacheItem = $this->createCacheItem($response);
+                $this->pool->save($cacheItem);
+            }
+
+            return $response;
+        });
+    }
+
+    /**
+     * We do only cache GET and HEAD requests.
+     *
+     * @param RequestInterface $request
+     *
+     * @return bool
+     */
+    private function isCacheableRequest(RequestInterface $request)
```

Comment:
Are you suggesting that I should rename `isCacheableRequest` to `isCacheable`. How can I make a difference between `isCacheableRequest` and `isCacheableResponse´?